pallets/uniques/src/lib.rs

```diff
@@ -0,0 +1,743 @@
+		///
+		/// Emits `CollectionMaxSupplySet` event when successful.
+		#[pallet::call_index(23)]
+		#[pallet::weight(<T as pallet_uniques::Config<I>>::WeightInfo::set_collection_max_supply())]
+		pub fn set_collection_max_supply(
+			origin: OriginFor<T>,
+			collection: T::CollectionId,
+			max_supply: u32,
+		) -> DispatchResult {
+			pallet_uniques::Pallet::<T, I>::set_collection_max_supply(origin, collection, max_supply)
+		}
+
+		/// Set (or reset) the price for an item.
+		///
+		/// Origin must be Signed and must be the owner of the asset `item`.
+		///
+		/// - `collection`: The collection of the item.
+		/// - `item`: The item to set the price for.
+		/// - `price`: The price for the item. Pass `None`, to reset the price.
+		/// - `buyer`: Restricts the buy operation to a specific account.
+		///
+		/// Emits `ItemPriceSet` on success if the price is not `None`.
+		/// Emits `ItemPriceRemoved` on success if the price is `None`.
+		#[pallet::call_index(24)]
+		#[pallet::weight(<T as pallet_uniques::Config<I>>::WeightInfo::set_price())]
+		pub fn set_price(
+			origin: OriginFor<T>,
+			collection: T::CollectionId,
+			item: T::ItemId,
+			price: Option<BalanceOf<T, I>>,
+			whitelisted_buyer: Option<AccountIdLookupOf<T>>,
+		) -> DispatchResult {
+			pallet_uniques::Pallet::<T, I>::set_price(origin, collection, item, price, whitelisted_buyer)
+		}
+
+		/// Allows to buy an item if it's up for sale.
+		///
+		/// Origin must be Signed and must not be the owner of the `item`.
+		///
+		/// - `collection`: The collection of the item.
+		/// - `item`: The item the sender wants to buy.
+		/// - `bid_price`: The price the sender is willing to pay.
+		///
+		/// Emits `ItemBought` on success.
+		#[pallet::call_index(25)]
+		#[pallet::weight(<T as pallet_uniques::Config<I>>::WeightInfo::buy_item())]
+		#[transactional]
+		pub fn buy_item(
+			origin: OriginFor<T>,
+			collection: T::CollectionId,
+			item: T::ItemId,
+			bid_price: BalanceOf<T, I>,
+		) -> DispatchResult {
+			pallet_uniques::Pallet::<T, I>::buy_item(origin, collection, item, bid_price)
+		}
+
+		/// Mint an item of a particular collection with extra deposit.
+		///
+		/// The origin must be Signed and the sender must be the Issuer of the `collection`.
+		///
+		/// - `collection`: The collection of the item to be minted.
+		/// - `item`: The item value of the item to be minted.
+		/// - `owner`: The initial owner of the minted item.
+		///
+		/// Emits `Issued` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[pallet::call_index(26)]
+		#[pallet::weight(<T as Config<I>>::WeightInfo::mint_with_extra_deposit())]
+		#[transactional]
+		pub fn mint_with_extra_deposit(
+			origin: OriginFor<T>,
+			collection: T::CollectionId,
+			item: T::ItemId,
+			owner: AccountIdLookupOf<T>,
+			deposit: BalanceOf<T, I>,
+		) -> DispatchResult {
+			pallet_uniques::Pallet::<T, I>::mint(origin, collection, item, owner).and_then(|_| {
+				let collection_owner =
```

Comment:
Should the reserve not be on owner, the beneficiary?

Reply:
If we want to go even further in making this operation secure is that the collection owner sets a maximum extra deposit that minter can use in total for minting the items of this collection.

Reply:
Probably the total_extra_deposit per collection by itself can do the job. When it's zero which is the default, no permission to mint with extra deposit.

Reply:
I like the idea of setting a cap and think this would be a light change for a higher level of robustness.

We definitely do not want to reserve tokens from the _item owner_ as this would open room for abuse and edge cases. It looks like it would make sense to keep the deposit in the collection owner account up until the NFT is burnt (which effectively leads to the item owner _claiming_ the deposit).

Reply:
Okay, I'd be implementing the cap shortly. CC @Hounsette

Reply:
Here is the commit that adds the discussed security feature: https://github.com/NodleCode/chain/pull/753/commits/9c6d8d888a6f111a671ecc093c0bb39a7901b1a6